### PR TITLE
refactor and expand results about Hellinger and Renyi divergences

### PR DESCRIPTION
**Pull request by LorenzoLuccioli**
- Add lemmas to Hellinger.lean
- Add section `MeasUnivAddMulHellingerDiv` with results about `ν(α) + (a - 1) * Hₐ(μ, ν)`
- Refactor lemmas to make the hypothesis `∀ᵐ x ∂μ, Integrable (fun b ↦ ((∂κ x/∂η x) b).toReal ^ a) (η x)` the standard one, instead of `∀ᵐ x ∂μ, Integrable (fun b ↦ hellingerFun a ((∂κ x/∂η x) b).toReal) (η x)`
- Add the DPI for the hHellinger divergence and related lemmas
- Change the definition of `renyiDiv` to use `EReal.log` instead of `Real.log` and adapt the API, in particular the conditions for the divergence to be infinite changed to `μ ⟂ₘ ν` in the case `a < 1`
- Add lemmas to Renyi.lean
- Remove the section `TopAndBounds` and add the section `RenyiEq`
- Add the DPI for the Renyi divergence and related lemmas
- Update the blueprint

Depends on #60, #61 and #64

### CODE REVIEW

**Comment by RemyDegenne** [comment]
Please merge master into this branch and resolve the conflicts.

File: TestingLowerBounds/Hellinger.lean [hunk 1]
```diff
@@ -100,6 +99,30 @@ lemma integrable_rpow_rnDeriv_iff [SigmaFinite ν] [SigmaFinite μ] (hμν : μ
     linarith
   · rw [rpow_add (ENNReal.toReal_pos h_zero hx), rpow_one]
 
+lemma integral_rpow_rnDeriv_eq_zero_iff_mutuallySingular [SigmaFinite μ] [SigmaFinite ν]
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean:111]:
```suggestion
-- TODO: generalize to the integral of any nonnegative function of the likelihood ratio
-- which is 0 only at 0.
lemma integral_rpow_rnDeriv_eq_zero_iff_mutuallySingular [SigmaFinite μ] [SigmaFinite ν]
```

File: TestingLowerBounds/Hellinger.lean [hunk 2]
```diff
@@ -100,6 +99,30 @@ lemma integrable_rpow_rnDeriv_iff [SigmaFinite ν] [SigmaFinite μ] (hμν : μ
...
+  have : 0 ≤ fun x ↦ ((∂μ/∂ν) x).toReal ^ a := by
+    intro x
+    simp only [Pi.zero_apply, ENNReal.toReal_nonneg, rpow_nonneg]
+  rw [integral_eq_zero_iff_of_nonneg this h_int]
+  apply Filter.eventually_congr
+  filter_upwards [Measure.rnDeriv_ne_top μ ν] with x hx
+  simp only [Pi.zero_apply, ENNReal.toReal_nonneg]
+  simp_rw [rpow_eq_zero ENNReal.toReal_nonneg ha_zero, ENNReal.toReal_eq_zero_iff, hx, or_false]
+
+lemma integral_rpow_rnDeriv_pos_iff_mutuallySingular [SigmaFinite μ] [SigmaFinite ν]
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]:
```suggestion
lemma integral_rpow_rnDeriv_pos_iff_not_mutuallySingular [SigmaFinite μ] [SigmaFinite ν]
```

File: TestingLowerBounds/Hellinger.lean [hunk 3]
```diff
@@ -100,6 +99,30 @@ lemma integrable_rpow_rnDeriv_iff [SigmaFinite ν] [SigmaFinite μ] (hμν : μ
...
+    (ha_zero : a ≠ 0) (h_int : Integrable (fun x ↦ ((∂μ/∂ν) x).toReal ^ a) ν) :
+    ∫ x, ((∂μ/∂ν) x).toReal ^ a ∂ν > 0 ↔ ¬ μ ⟂ₘ ν := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]:
```suggestion
    0 < ∫ x, ((∂μ/∂ν) x).toReal ^ a ∂ν ↔ ¬ μ ⟂ₘ ν := by
```

File: TestingLowerBounds/Hellinger.lean [hunk 4]
```diff
@@ -153,13 +176,20 @@ lemma stronglyMeasurable_hellingerFun (ha_nonneg : 0 ≤ a) :
     measurability
 
 @[simp]
-lemma hellingerFun_one_eq_zero : hellingerFun a 1 = 0 := by
+lemma hellingerFun_at_one_eq_zero : hellingerFun a 1 = 0 := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]: `at` is not one of the words mathlib uses in names.
```suggestion
lemma hellingerFun_apply_one_eq_zero : hellingerFun a 1 = 0 := by
```

File: TestingLowerBounds/Hellinger.lean [hunk 5]
```diff
@@ -153,13 +176,20 @@ lemma stronglyMeasurable_hellingerFun (ha_nonneg : 0 ≤ a) :
...
   by_cases ha_one : a = 1
   · simp [ha_one, hellingerFun_one]
   by_cases ha_zero : a = 0
   · simp [ha_zero, hellingerFun_zero]
   simp [hellingerFun, ha_one, ha_zero]
 
+lemma hellingerFun_at_zero : hellingerFun a 0 = (1 - a)⁻¹ := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]:
```suggestion
lemma hellingerFun_apply_zero : hellingerFun a 0 = (1 - a)⁻¹ := by
```

File: TestingLowerBounds/Hellinger.lean [hunk 6]
```diff
@@ -311,6 +341,23 @@ lemma hellingerDiv_zero_ne_top (μ ν : Measure α) [IsFiniteMeasure ν] :
     hellingerDiv 1 μ ν = kl μ ν := by
   rw [hellingerDiv, hellingerFun_one, kl_eq_fDiv]
 
+@[simp]
+lemma hellingerDiv_zero_measure (ν : Measure α) [IsFiniteMeasure ν] :
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]:
```suggestion
lemma hellingerDiv_zero_measure_left (ν : Measure α) [IsFiniteMeasure ν] :
```

File: TestingLowerBounds/Hellinger.lean [hunk 7]
```diff
@@ -498,23 +578,189 @@ lemma hellingerDiv_nonneg (ha_pos : 0 ≤ a) (μ ν : Measure α)
...
+lemma meas_univ_add_mul_hellingerDiv_zero_eq (ha : a = 0) [IsFiniteMeasure μ] [IsFiniteMeasure ν] :
+    ↑(ν Set.univ) + (a - 1) * hellingerDiv a μ ν = ν {x | 0 < (∂μ/∂ν) x} := by
+  simp only [ha, EReal.coe_zero, zero_sub, hellingerDiv_zero'', neg_mul, one_mul, rpow_zero,
+    integral_const, smul_eq_mul, mul_one]
+  rw [EReal.neg_sub, ← add_assoc, ← sub_eq_add_neg, EReal.sub_self, zero_add]
+  all_goals simp [measure_ne_top]
+
+lemma meas_univ_add_mul_hellingerDiv_nonneg_of_le_one (ha_nonneg : 0 ≤ a) (ha : a ≤ 1)
+    (μ ν : Measure α) [IsFiniteMeasure μ] [IsFiniteMeasure ν] :
+    ↑(ν Set.univ) + (a - 1) * hellingerDiv a μ ν ≥ 0 := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]:
```suggestion
    0 ≤ ↑(ν Set.univ) + (a - 1) * hellingerDiv a μ ν := by
```

File: TestingLowerBounds/Hellinger.lean [hunk 8]
```diff
@@ -498,23 +578,189 @@ lemma hellingerDiv_nonneg (ha_pos : 0 ≤ a) (μ ν : Measure α)
...
+    _ = (ν Set.univ) - (ν Set.univ) := by
+      norm_cast
+      rw [← mul_assoc, ← EReal.coe_mul, mul_inv_cancel (by linarith), EReal.coe_one, one_mul]
+    _ ≥ _ := by
+      rw [← ENNReal.toEReal_sub (measure_ne_top _ _) (le_refl _)]
+      simp
+
+lemma meas_univ_add_mul_hellingerDiv_nonneg_of_one_lt (ha : 1 < a) (μ ν : Measure α)
+    [IsFiniteMeasure μ] [IsFiniteMeasure ν] :
+    ↑(ν Set.univ) + (a - 1) * hellingerDiv a μ ν ≥ 0 := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]:
```suggestion
    0 ≤ ↑(ν Set.univ) + (a - 1) * hellingerDiv a μ ν := by
```

File: TestingLowerBounds/Hellinger.lean [hunk 9]
```diff
@@ -498,23 +578,189 @@ lemma hellingerDiv_nonneg (ha_pos : 0 ≤ a) (μ ν : Measure α)
...
+    · exact OrderTop.le_top 0
+    · norm_cast
+      linarith
+  rw [meas_univ_add_mul_hellingerDiv_eq (by linarith) ha.ne' h_top]
+  simp only [ge_iff_le, EReal.coe_nonneg]
+  positivity
+
+lemma meas_univ_add_mul_hellingerDiv_nonneg (ha_nonneg : 0 ≤ a) (μ ν : Measure α)
+    [IsFiniteMeasure μ] [IsFiniteMeasure ν] :
+    ↑(ν Set.univ) + (a - 1) * hellingerDiv a μ ν ≥ 0 := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/Hellinger.lean]:
```suggestion
    0 ≤ ↑(ν Set.univ) + (a - 1) * hellingerDiv a μ ν := by
```

File: TestingLowerBounds/Renyi.lean [hunk 10]
```diff
@@ -51,203 +53,266 @@ lemma exp_mul_llr' [SigmaFinite μ] [SigmaFinite ν] (hμν : μ ≪ ν) :
...
 `(a - 1)⁻¹ * log (1 + (a - 1) * Hₐ(μ, ν))`, but this definition maintains some useful properties
 also for a general finite measure `ν`, in particular the integral form
-`Rₐ(μ, ν) = (a - 1)⁻¹ * log (∫ x, ((∂μ/∂ν) x).toReal ^ a ∂ν)`. -/
+`Rₐ(μ, ν) = (a - 1)⁻¹ * log (∫ x, ((∂μ/∂ν) x) ^ a ∂ν)`. -/
 noncomputable def renyiDiv (a : ℝ) (μ ν : Measure α) : EReal :=
   if a = 1 then kl μ ν
-  else if hellingerDiv a μ ν ≠ ⊤
-    then (a - 1)⁻¹ * log ((ν Set.univ).toReal + (a - 1) * (hellingerDiv a μ ν).toReal)
-    else ⊤
+  else (a - 1)⁻¹ * EReal.log ((↑(ν Set.univ) + (a - 1) * (hellingerDiv a μ ν)).toENNReal)
```
Comment (RemyDegenne) [inline TestingLowerBounds/Renyi.lean:83]: Can you please write something in the file docstring (in an "implementation notes" section), or in the docstring of this def, about the use of EReal.log in this definition?

File: TestingLowerBounds/Renyi.lean [hunk 11]
```diff
@@ -51,203 +53,266 @@ lemma exp_mul_llr' [SigmaFinite μ] [SigmaFinite ν] (hμν : μ ≪ ν) :
...
+      = (a - 1)⁻¹ * EReal.log ((↑(ν Set.univ) + (a - 1) * (hellingerDiv a μ ν)).toENNReal) := by
+  rw [renyiDiv, if_neg ha_ne_one]
 
-lemma renyiDiv_eq_top_iff_hellingerDiv_eq_top' (ha_ne_one : a ≠ 1) :
-    renyiDiv a μ ν = ⊤ ↔ hellingerDiv a μ ν = ⊤ := by
-  simp only [renyiDiv, ha_ne_one, ↓reduceIte, ne_eq, ite_not, ite_eq_left_iff]
-  rw [← EReal.coe_mul]
-  simp only [EReal.coe_ne_top, imp_false, not_not]
+@[simp]
+lemma renyiDiv_zero_measure (ν : Measure α) [IsFiniteMeasure ν] :
```
Comment (RemyDegenne) [inline TestingLowerBounds/Renyi.lean]:
```suggestion
lemma renyiDiv_zero_measure_left (ν : Measure α) [IsFiniteMeasure ν] :
```

File: TestingLowerBounds/Renyi.lean [hunk 12]
```diff
@@ -51,203 +53,266 @@ lemma exp_mul_llr' [SigmaFinite μ] [SigmaFinite ν] (hμν : μ ≪ ν) :
...
 
-lemma renyiDiv_of_not_integrable' (ha_ne_one : a ≠ 1)
-    (h_int : ¬ Integrable (fun x ↦ hellingerFun a ((∂μ/∂ν) x).toReal) ν) :
+lemma renyiDiv_eq_top_iff_mutuallySingular_of_lt_one (ha_nonneg : 0 ≤ a) (ha : a < 1)
+    [IsFiniteMeasure μ] [IsFiniteMeasure ν] :
+    renyiDiv a μ ν = ⊤ ↔ μ ⟂ₘ ν := by
+  rw [renyiDiv_of_ne_one ha.ne, EReal.mul_eq_top,
+    ← toENNReal_meas_univ_add_mul_hellingerDiv_eq_zero_iff_of_lt_one ha_nonneg ha]
+  simp [ha, not_lt_of_gt ha]
+--offspring
```
Comment (RemyDegenne) [inline TestingLowerBounds/Renyi.lean]:
```suggestion

```

File: TestingLowerBounds/Renyi.lean [hunk 13]
```diff
@@ -51,203 +53,266 @@ lemma exp_mul_llr' [SigmaFinite μ] [SigmaFinite ν] (hμν : μ ≪ ν) :
...
-
-lemma renyiDiv_of_one_lt_of_integrable_of_ac (ha_one_lt : 1 < a) [IsFiniteMeasure μ] [SigmaFinite ν]
-    (h_int : Integrable (fun x ↦ hellingerFun a ((∂μ/∂ν) x).toReal) ν) (hμν : μ ≪ ν) :
-    renyiDiv a μ ν
-      = (a - 1)⁻¹ * log ((ν Set.univ).toReal + (a - 1) * (hellingerDiv a μ ν).toReal) := by
-  rw [renyiDiv, if_neg ha_one_lt.ne',
-    if_pos ((hellingerDiv_ne_top_iff_of_one_le ha_one_lt.le _ _).mpr ⟨h_int, hμν⟩)]
-
-end TopAndBounds
+lemma renyiDiv_eq_top_forall_of_eq_top_of_lt_one (ha_nonneg : 0 ≤ a) (ha : a < 1) [NeZero μ]
```
Comment (RemyDegenne) [inline TestingLowerBounds/Renyi.lean]:
```suggestion
lemma forall_renyiDiv_eq_top_of_eq_top_of_lt_one (ha_nonneg : 0 ≤ a) (ha : a < 1) [NeZero μ]
```

File: TestingLowerBounds/Renyi.lean [hunk 14]
```diff
@@ -274,54 +336,66 @@ lemma renyiDiv_symm (ha_pos : 0 < a) (ha : a < 1)
...
+    exact hν.out <| Measure.eq_zero_of_absolutelyContinuous_of_mutuallySingular hμν h
+  rw [EReal.log_ofReal_of_pos]
+  swap;
+  · rw [← integral_rnDeriv_smul hμν]
+    simp_rw [smul_eq_mul, mul_comm ((∂μ/∂ν) _).toReal,
+      ← Real.rpow_add_one' ENNReal.toReal_nonneg (by linarith), add_comm a]
+    exact integral_rpow_rnDeriv_pos_iff_mutuallySingular (by linarith) h_int |>.mpr h_ms
   congr 2
-  exact integral_congr_ae (exp_mul_llr' hμν)
+  refine integral_congr_ae (exp_mul_llr' hμν)
```
Comment (RemyDegenne) [inline TestingLowerBounds/Renyi.lean]:
```suggestion
  exact integral_congr_ae (exp_mul_llr' hμν)
```

File: TestingLowerBounds/Renyi.lean [hunk 15]
```diff
@@ -274,54 +336,66 @@ lemma renyiDiv_symm (ha_pos : 0 < a) (ha : a < 1)
...
+
```
Comment (RemyDegenne) [inline TestingLowerBounds/Renyi.lean]:
```suggestion
```

File: TestingLowerBounds/Hellinger.lean (hunk 1 above)
Comment (LorenzoLuccioli) [reply TestingLowerBounds/Hellinger.lean:111]: I directly added the generalized lemma, was this what you had in mind? 
I was not sure what should be the domain of the function. I used `f : ENNReal → ℝ`, in this way it's easier to state the nonnegativity and we don't have to compose with `.toReal`, but on the other hand we need to handle the case where the argument is infinte, so the other hypothesis becomes: `∀ x, f x = 0 ↔ x = 0 ∨ x = ⊤`.

Comment (LorenzoLuccioli) [reply TestingLowerBounds/Hellinger.lean:111]: Also the name of the new lemma probably needs to be changed.

**Review by RemyDegenne — APPROVED** [review]

**Comment by RemyDegenne** [comment]
LGTM!
Please resolve the conflict then merge.

**Pull request merged**